docs/zkapps/tutorials/04-in-browser-react-ui.mdx

```diff
@@ -0,0 +1,407 @@
+
+The second shows typescript errors. You can watch this as you develop to check for type errors.
+
+### Implementing the react app
+
+Open up `ui/pages/_app.page.tsx` in your editor. You can find the full contents for this file [here](https://raw.githubusercontent.com/es92/zkApp-examples/main/04-zkapp-browser-ui/ui/pages/_app.page.tsx) for reference.
+
+Edit the contents so it looks like this:
+```ts
+  1 import '../styles/globals.css'
+  2 import React, { useEffect, useState } from "react";
+  3 import './reactCOIServiceWorker';
+  4
+  5 import ZkappWorkerClient from './zkappWorkerClient';
+  6
+  7 import {
+  8   PublicKey,
+  9   PrivateKey,
+ 10   Field,
+ 11 } from 'snarkyjs'
+ 12
+ 13 let transactionFee = 100_000_000;
+ 14
+ 15 export default function App() {
+ 16   return </div>
+ 17 }
+ 18
+```
+
+This just sets up our react project, with the imports we'll need, and an empty component.
+
+Now, we'll add state to our app:
+
+```ts
+...
+ 15 export default function App() {
+ 16
+ 17   let [state, setState] = useState({
+ 18     zkappWorkerClient: null as null | ZkappWorkerClient,
+ 19     hasBeenSetup: false,
+ 20     accountExists: false,
+ 21     currentNum: null as null | Field,
+ 22     privateKey: null as null | PrivateKey,
+ 23     publicKey: null as null | PublicKey,
+ 24     zkappPublicKey: null as null | PublicKey,
+ 25     creatingTransaction: false,
+ 26   });
+ 27
+ 28   // -------------------------------------------------------
+ 29
+ 30   return </div>
+...
+```
+
+This creates mutable state that we can reference in our UI, and update as our application runs. You can learn more about react state and useState [here](https://reactjs.org/docs/hooks-state.html).
+
+Next, we'll add a function to setup our application:
+```ts
+...
+ 28   // -------------------------------------------------------
+ 29   // Do Setup
+ 30
+ 31   useEffect(() => {
+ 32     (async () => {
+ 33       if (!state.hasBeenSetup) {
+ 34         const zkappWorkerClient = new ZkappWorkerClient();
+ 35
+ 36         console.log('Loading SnarkyJS...');
+ 37         await zkappWorkerClient.loadSnarkyJS();
+ 38         console.log('done');
+ 39
+ 40         await zkappWorkerClient.setActiveInstanceToBerkeley();
+ 41
+ 42         // TODO
+ 43       }
+ 44     })();
+ 45   }, []);
+ 46
+ 47   // -------------------------------------------------------
+...
+```
+
+
+We use the react feature "useEffect", so that this is only run once. See more on useEffect [here](https://reactjs.org/docs/hooks-effect.html). We further gate running this effect by a boolean `hasBeenSetup` so we don't run this more than once.
+
+This is also where we setup our web worker client. This will interact with our web worker running snarkyjs code, so that that code doesn't block the UI thread.
+
+Next, we check if there is already a private key in local storage - and if not, we save one there. We also check if the account exists on chain:
```

Comment:
should there be a sentence / small disclaimer here, that storing user's private keys in localstorage is not good enough for production apps?
(I built an app that did this once, and had to realize that some users have their browsers set up so that localstorage gets cleared out periodically...)

Reply:
Yeah definitely. Added cautions in the two places where localStorage is mentioned

Reply:
oh I missed the first one. sgtm!

Reply:
To be refactored, for Auro Wallet instructions.